lib/srv/discovery/fetchers/kube_services.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fetchers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/discovery/common"
+)
+
+// KubeAppsFetcherConfig configures kubeAppFetcher
+type KubeAppsFetcherConfig struct {
+	// Name of the kubernetes cluster
+	ClusterName string
+	// KubernetesClient is a client for Kubernetes API
+	KubernetesClient kubernetes.Interface
+	// FilterLabels are the filter criteria.
+	FilterLabels types.Labels
+	// Namespaces are the kubernetes namespaces in which to discover services
+	Namespaces []string
+	// Log is a logger to use
+	Log logrus.FieldLogger
+	// PI inspects port to find your whether they are HTTP/HTTPS or not.
+	protocolChecker services.ProtocolChecker
+}
+
+// CheckAndSetDefaults validates and sets the defaults values.
+func (k *KubeAppsFetcherConfig) CheckAndSetDefaults() error {
+	if k.FilterLabels == nil {
+		return trace.BadParameter("missing parameter AccessPoint")
+	}
+	if k.KubernetesClient == nil {
+		return trace.BadParameter("missing parameter KubernetesClient")
+	}
+	if k.Log == nil {
+		return trace.BadParameter("missing parameter Log")
+	}
+	if k.ClusterName == "" {
+		return trace.BadParameter("missing parameter ClusterName")
+	}
+	if k.protocolChecker == nil {
+		k.protocolChecker = &noopProtocolChecker{}
+	}
+
+	return nil
+}
+
+// kubeAppFetcher fetches app resources from Kubernetes services
+type kubeAppFetcher struct {
+	KubeAppsFetcherConfig
+}
+
+// Default implementation, doesn't actually performs HTTP request.
+type noopProtocolChecker struct{}
+
+// CheckProtocol for noopProtocolChecker just returns 'tcp'
+func (*noopProtocolChecker) CheckProtocol(uri string) string {
+	return "tcp"
+}
+
+// NewKubeAppsFetcher creates new Kubernetes app fetcher
+func NewKubeAppsFetcher(cfg KubeAppsFetcherConfig) (common.Fetcher, error) {
```

Comment:
Right, changed it to return concrete type. I returned `common.Fetcher` to be consistent, since all other matchers return it, but I agree that this is against best practice.